keps/sig-api-machinery/3156-http3/README.md

```diff
@@ -0,0 +1,228 @@
+simplifies the implementation, a working prototype can be seen here https://github.com/kubernetes/kubernetes/pull/106707/.
+
+### kube-apiserver
+
+
+A new UDP listener should be added to the kube-apiserver in the same port and address that the current TCP listener.
+
+```go
+// https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/staging/src/k8s.io/apiserver/pkg/server/secure_serving.go#L276
+func RunServer(
+	server *http.Server,
+	ln net.Listener,
+	shutDownTimeout time.Duration,
+	stopCh <-chan struct{},
+) (<-chan struct{}, <-chan struct{}, error) {
+```
+
+In addition to the listener, there is some logic inside the apiserver that discriminates by protocol like `WrapForHTTP1Or2` that should be adapted.
+
+```go
+// https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go#L58
+// WrapForHTTP1Or2 accepts a user-provided decorator of an "inner" http.responseWriter
+// object and potentially wraps the user-provided decorator with a new http.ResponseWriter
+// object that implements http.CloseNotifier, http.Flusher, and/or http.Hijacker by
+// delegating to the user-provided decorator (if it implements the relevant method) or
+// the inner http.ResponseWriter (otherwise), so that the returned http.ResponseWriter
+// object implements the same subset of those interfaces as the inner http.ResponseWriter.
+...
+func WrapForHTTP1Or2(decorator UserProvidedDecorator) http.ResponseWriter {
+```
+
+### client-go
+
+The golang http/3 implementation exposes a Transport roundtripper, there are two options to use it in client-go:
+
+1. Use the WrapTransport option and let users configure the roundtripper manually.
+
+```go
+  tlsConfig, err := TLSConfigFor(config)
+  if err != nil {
+    return nil, err
+  }
+  rt = &http3.RoundTripper{
+    TLSClientConfig: tlsConfig,
+    QuicConfig: &quic.Config{
+      KeepAlive: true,
+    },
+  }
+
+config.WrapTransport = rt
+```
+
+
+2. Add a new option to the RESTConfig of client-go to enable http3 and automate the HTTP3 roundtripper configuration:
+
+```go
+// staging/src/k8s.io/client-go/transport/config.go
+// Config holds various options for establishing a transport.
+type Config struct {
+	// UserAgent is an optional field that specifies the caller of this
+	// request.
+	UserAgent string
+	// The base TLS configuration for this transport.
+	TLS TLSConfig
+
+// Use HTTP3
+	EnableHTTP3 bool
+```
+
+### Test Plan
+
+All current tests should pass using the new protocol, this can be done by enabling by default http3 in client go so all the
+components of the cluster switch to it.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented in both apiserver and client-go behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- HTTP/3 becomes an Internet RFC
```

Comment:
I think we could graduate the feature to beta before the RFC, provided specifically that the feature gate is off by default until the RFC is out.